File: Dynamic Programming/45. Jump Game II/solutionn.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        memo = {}
        def solve(index):
            if index in memo:
                return memo[index]
            if index==len(nums)-1:
                return 0
            ans = math.inf
            if nums[index]>0:
                for i in range(1,nums[index]+1):
                    if i+index<len(nums):
                        ans = min(ans,1+solve(index+i))
                    else:
                        break
            memo[index]=ans
            return memo[index]
        return solve(0)

File: Dynamic Programming/746. Min Cost Climbing Stairs/Solutionn.java
class Solution {

    private Map<Integer, Integer> memo = new HashMap<>();

    public int minCostClimbingStairs(int[] cost) {
        return Math.min(this.solve(0, cost), this.solve(1, cost));
    }

    private int solve(int idx, int[] cost) {
        if (this.memo.containsKey(idx)) {
            return this.memo.get(idx);
        }
        if (idx == cost.length) {
            return 0;
        }
        if (idx == cost.length - 1) {
            this.memo.put(idx, cost[idx]);
            return cost[idx];
        }

        int oneStep = this.solve(idx + 1, cost);
        int twoStep = this.solve(idx + 2, cost);
        int ans = Math.min(oneStep, twoStep);
        this.memo.put(idx, cost[idx] + ans);
        return this.memo.get(idx);
    }
}